### ISSUE TITLE
Flaky `test_pay_disconnect_stress` due to capacity exhaustion

### ISSUE THREAD

**Issue by cdecker** [ElementsProject/lightning#3649]
`test_pay_disconnect_stress` fails due to an `update_add_htlc` that exhausts the available channel capacity. The sender should likely not even attempt to add that HTLC:

```log
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.305Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-channeld-chan#1: peer_out WIRE_UPDATE_ADD_HTLC
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.305Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-channeld-chan#1: peer_out WIRE_UPDATE_FULFILL_HTLC
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.306Z DEBUG 0265b6ab5ec860cd257865d61ef0bbf5b3339c36cbda8b26b74e7f1dca490b6518-gossipd: Attempted to local_add_channel a known channel
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.306Z DEBUG gossipd: channel_update 107x1x1/1: delaying 2 secs
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.306Z DEBUG 0265b6ab5ec860cd257865d61ef0bbf5b3339c36cbda8b26b74e7f1dca490b6518-channeld-chan#1: peer_in WIRE_UPDATE_ADD_HTLC
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.306Z DEBUG 0265b6ab5ec860cd257865d61ef0bbf5b3339c36cbda8b26b74e7f1dca490b6518-channeld-chan#1: Failed to add 1 remove 0 htlcs
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.306Z DEBUG 0265b6ab5ec860cd257865d61ef0bbf5b3339c36cbda8b26b74e7f1dca490b6518-channeld-chan#1: peer_out WIRE_ERROR
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.306Z UNUSUAL 0265b6ab5ec860cd257865d61ef0bbf5b3339c36cbda8b26b74e7f1dca490b6518-chan#1: Peer permanent failure in CHANNELD_NORMAL: channeld: sent ERROR Bad peer_add_htlc: CHANNEL_ERR_CHANNEL_CAPACITY_EXCEEDED
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.308Z DEBUG lightningd: Broadcasting txid 79175f3b295c2c43b8a2919d08333dda393241378ab568757bd0776b78d5761a
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.308Z DEBUG lightningd: sendrawtransaction: 02000000000101bbfcd981b7366fe106090114dc1db0d401779b3198e0c8e2f3ec15920fc324eb0100000000b637118002a06c07000000000022002030334aa481d2b33a88da94c8efa27e7f284243afce475fe944123d9271c9e67920a1070000000000220020c9a8bdaca7f11be0c440775a87f2b910572a19275a7c0654eba2f0313ce851f004004730440220179df737e6c9f709dc72522ec9b0d8e46b809d590f56a6980a856827653cc7030220781cc9195fb163d9901459746014c61244994f836cc1ffbd6bdc73be458d3b940147304402200b8d2ef6021114f837b20b90d8d8d7671d622f5c68460f948a5328e40d29c9ad02204e8a9fbb3491c3285fd2d572a1adb09ad61cc990dde3f7581a3376cf5c2aa5eb014752210362a03b58ce471d2bd4e1c36ba06944e6952a485929b25958a38fc4cfe0f457632103dce0e1c5b449fc4deb8dfd3f4ae99be642b32b1744134de520925ed14df803dc52ae7d5fd220
DEBUG:root:lightningd-5: 2020-04-16T08:47:59.308Z INFO 0265b6ab5ec860cd257865d61ef0bbf5b3339c36cbda8b26b74e7f1dca490b6518-chan#1: State changed from CHANNELD_NORMAL to AWAITING_UNILATERAL
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.307Z DEBUG hsmd: Client: Received message 19 from client
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.307Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-gossipd: Attempted to local_add_channel a known channel
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.307Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-channeld-chan#1: Creating commit_sig signature 2 304402202634ca5c11b32f3327731fa3cdc68155b03f3bcae0ce4de2130e5f66127bfe170220701a04578e42422cd8e40bb880688b7504376ed5eb67ae7f905b6abda6b3f17a01 for tx 0200000001bbfcd981b7366fe106090114dc1db0d401779b3198e0c8e2f3ec15920fc324eb0100000000b637118002b476070000000000220020bba12b99ee09a4a1c835f51dc20c62c9e90f6b4a66222efc806501d6a5c4d5d116a1070000000000160014cb54d14a66e321b312aee57a694518d4fd9656f77e5fd220 wscript 52210362a03b58ce471d2bd4e1c36ba06944e6952a485929b25958a38fc4cfe0f457632103dce0e1c5b449fc4deb8dfd3f4ae99be642b32b1744134de520925ed14df803dc52ae key 03dce0e1c5b449fc4deb8dfd3f4ae99be642b32b1744134de520925ed14df803dc
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.307Z DEBUG gossipd: channel_update 107x1x1/0: delaying 2 secs
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.307Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-channeld-chan#1: peer_out WIRE_COMMITMENT_SIGNED
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.308Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-channeld-chan#1: peer_in WIRE_ERROR
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.309Z INFO 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-chan#1: Peer transient failure in CHANNELD_NORMAL: channeld: (ignoring) channel bbfcd981b7366fe106090114dc1db0d401779b3198e0c8e2f3ec15920fc324ea: Bad peer_add_htlc: CHANNEL_ERR_CHANNEL_CAPACITY_EXCEEDED
DEBUG:root:lightningd-6: 2020-04-16T08:47:59.309Z DEBUG 032cf15d1ad9c4a08d26eab1918f732d8ef8fdc6abb9640bf3db174372c491304e-chan#1: Will try reconnect in 60 seconds
```

After this `lightningd-6` just sits there and waits for `lightningd-5` to
accept its reestablish attempt, which it never will...

Currently bisecting `master` to see where the issue started happening.

**Comment by cdecker** [comment]
Bisecting it seems that 7a95e90ee4d4143ffd81814587433a8490a59c2b is the first failing commit, however I fail to see how that could be the case. I ran that commit and its parent 500 times through the failing test, and the results are consistent, but that commit doesn't change any of the HTLC logic, just some feature signaling. Maybe @rustyrussell has an idea?

**Comment by niftynei** [comment]
This has to do with how we verify capacity before accepting incoming HTLCs; the problem is triggered when we have an outstanding HTLC to fulfill and an outstanding HTLC to add at reconnect. If the add request is sent before the fulfill, then the peer fails with the `CAPACITY_EXCEEDED` and closes the channel, which caused the test to timeout as the offending peer waits forever for its payments to resolve.

Fixed in https://github.com/ElementsProject/lightning/pull/3650/commits/a713c3f23991b3c1deab459cdd4444977e2e2050 by always sending `fulfill`s before `adds` for htlcs on reconnect (and shipped in the 0.8.1rc1).
